File: src/world/level/chunk/data-layer.ts
export class DataLayer {
  public static readonly LAYER_COUNT = 16;
  public static readonly LAYER_SIZE = 128;
  public static readonly SIZE = 2048;

  private data: Uint8Array | undefined;

  public constructor(data?: Uint8Array) {
    if (data !== undefined && data.length !== DataLayer.SIZE) {
      throw new Error(`DataLayer should be ${DataLayer.SIZE} bytes not: ${data.length}`);
    }
    this.data = data;
  }

  public get(x: number, y: number, z: number): number {
    return this.getByIndex(DataLayer.getIndex(x, y, z));
  }

  public set(x: number, y: number, z: number, value: number): void {
    this.setByIndex(DataLayer.getIndex(x, y, z), value);
  }

  private static getIndex(x: number, y: number, z: number): number {
    return y << 8 | z << 4 | x;
  }

  private getByIndex(index: number): number {
    if (this.data === undefined) {
      return 0;
    }

    const byteIndex = DataLayer.getByteIndex(index);
    const nibbleIndex = DataLayer.getNibbleIndex(index);
    return this.data[byteIndex]! >> (4 * nibbleIndex) & 15;
  }

  private setByIndex(index: number, value: number): void {
    if (this.data === undefined) {
      this.data = new Uint8Array(DataLayer.SIZE);
    }

    const byteIndex = DataLayer.getByteIndex(index);
    const nibbleIndex = DataLayer.getNibbleIndex(index);
    const clearMask = ~(15 << (4 * nibbleIndex));
    const valueBits = (value & 15) << (4 * nibbleIndex);
    this.data[byteIndex] = (this.data[byteIndex]! & clearMask) | valueBits;
  }

  private static getNibbleIndex(index: number): number {
    return index & 1;
  }

  private static getByteIndex(index: number): number {
    return index >> 1;
  }

  public getData(): Uint8Array {
    if (this.data === undefined) {
      this.data = new Uint8Array(DataLayer.SIZE);
    }
    return this.data;
  }

  public copy(): DataLayer {
    return this.data === undefined ? new DataLayer() : new DataLayer(new Uint8Array(this.data));
  }

  public toString(): string {
    let out = "";
    for (let index = 0; index < 4096; index++) {
      out += this.getByIndex(index).toString(16);
      if ((index & 15) === 15) {
        out += "\n";
      }
      if ((index & 0xFF) === 255) {
        out += "\n";
      }
    }
    return out;
  }

  public layerToString(_layer: number): string {
    let out = "";
    for (let index = 0; index < 256; index++) {
      out += this.getByIndex(index).toString(16);
      if ((index & 15) === 15) {
        out += "\n";
      }
    }
    return out;
  }

  public isEmpty(): boolean {
    return this.data === undefined;
  }
}

File: native/crates/mclone-light/src/data_layer.rs
use std::fmt;

use mclone_core::LIGHT_DATA_LAYER_BYTE_COUNT;

pub const DATA_LAYER_SIZE: usize = LIGHT_DATA_LAYER_BYTE_COUNT;
pub const DATA_LAYER_VALUE_COUNT: usize = 4096;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DataLayer {
    data: Option<Box<[u8; DATA_LAYER_SIZE]>>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum DataLayerError {
    InvalidByteLength { actual: usize },
}

impl fmt::Display for DataLayerError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidByteLength { actual } => write!(
                f,
                "DataLayer should be {DATA_LAYER_SIZE} bytes, got {actual}"
            ),
        }
    }
}

impl std::error::Error for DataLayerError {}

impl Default for DataLayer {
    fn default() -> Self {
        Self::new()
    }
}

impl DataLayer {
    pub fn new() -> Self {
        Self { data: None }
    }

    pub fn from_bytes(bytes: [u8; DATA_LAYER_SIZE]) -> Self {
        Self {
            data: Some(Box::new(bytes)),
        }
    }

    pub fn from_vec(bytes: Vec<u8>) -> Result<Self, DataLayerError> {
        let actual = bytes.len();
        let bytes: [u8; DATA_LAYER_SIZE] = bytes
            .try_into()
            .map_err(|_| DataLayerError::InvalidByteLength { actual })?;
        Ok(Self::from_bytes(bytes))
    }

    pub fn get(&self, x: i32, y: i32, z: i32) -> u8 {
        self.get_index(data_layer_index(x, y, z))
    }

    pub fn set(&mut self, x: i32, y: i32, z: i32, value: u8) {
        self.set_index(data_layer_index(x, y, z), value);
    }

    pub fn get_index(&self, index: usize) -> u8 {
        assert!(
            index < DATA_LAYER_VALUE_COUNT,
            "DataLayer index {index} out of bounds"
        );
        let Some(data) = &self.data else {
            return 0;
        };
        let byte = data[index >> 1];
        let shift = 4 * (index & 1);
        (byte >> shift) & 15
    }

    pub fn set_index(&mut self, index: usize, value: u8) {
        assert!(
            index < DATA_LAYER_VALUE_COUNT,
            "DataLayer index {index} out of bounds"
        );
        let data = self
            .data
            .get_or_insert_with(|| Box::new([0; DATA_LAYER_SIZE]));
        let byte_index = index >> 1;
        let shift = 4 * (index & 1);
        let mask = !(15 << shift);
        data[byte_index] = (data[byte_index] & mask) | ((value & 15) << shift);
    }

    pub fn get_data(&mut self) -> &[u8; DATA_LAYER_SIZE] {
        self.data
            .get_or_insert_with(|| Box::new([0; DATA_LAYER_SIZE]))
    }

    pub fn get_data_mut(&mut self) -> &mut [u8; DATA_LAYER_SIZE] {
        self.data
            .get_or_insert_with(|| Box::new([0; DATA_LAYER_SIZE]))
    }

    pub fn as_bytes(&self) -> Option<&[u8; DATA_LAYER_SIZE]> {
        self.data.as_deref()
    }

    pub fn into_bytes(self) -> Option<Vec<u8>> {
        self.data.map(|data| Vec::from(*data))
    }

    pub fn is_empty(&self) -> bool {
        self.data.is_none()
    }
}

pub fn data_layer_index(x: i32, y: i32, z: i32) -> usize {
    assert!((0..16).contains(&x), "DataLayer x {x} out of bounds");
    assert!((0..16).contains(&y), "DataLayer y {y} out of bounds");
    assert!((0..16).contains(&z), "DataLayer z {z} out of bounds");
    ((y << 8) | (z << 4) | x) as usize
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn data_layer_empty_reads_zero_and_allocates_lazily() {
        let mut layer = DataLayer::new();

        assert!(layer.is_empty());
        assert_eq!(layer.get(15, 15, 15), 0);
        assert!(layer.is_empty());
        assert_eq!(layer.get_data().len(), DATA_LAYER_SIZE);
        assert!(!layer.is_empty());
    }

    #[test]
    fn data_layer_mut_data_allocates_and_allows_bulk_fill() {
        let mut layer = DataLayer::new();

        layer.get_data_mut().fill(0xFF);

        assert!(!layer.is_empty());
        assert_eq!(layer.get(0, 0, 0), 15);
        assert_eq!(layer.get(15, 15, 15), 15);
    }

    #[test]
    fn data_layer_uses_java_index_and_nibble_order() {
        let mut layer = DataLayer::new();

        layer.set(0, 0, 0, 0x0A);
        layer.set(1, 0, 0, 0x03);
        layer.set(15, 15, 15, 0x0F);

        let data = layer.as_bytes().unwrap();
        assert_eq!(data[0], 0x3A);
        assert_eq!(data_layer_index(0, 1, 0), 256);
        assert_eq!(data_layer_index(15, 15, 15), 4095);
        assert_eq!(layer.get(0, 0, 0), 0x0A);
        assert_eq!(layer.get(1, 0, 0), 0x03);
        assert_eq!(layer.get(15, 15, 15), 0x0F);
    }

    #[test]
    fn data_layer_copy_is_independent() {
        let mut original = DataLayer::new();
        original.set(2, 3, 4, 7);
        let mut copied = original.clone();

        copied.set(2, 3, 4, 12);

        assert_eq!(original.get(2, 3, 4), 7);
        assert_eq!(copied.get(2, 3, 4), 12);
    }
}
